core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -116,174 +118,179 @@ sealed trait Channel[F[_], A] {
+                        else
+                          Applicative[F].unit
                       }
-
-                      val toEmit = makeChunk(initValues, tailValues.result(), size)
-
-                      unblock.as(Pull.output(toEmit) >> consumeLoop)
                     } else {
-                      F.pure(
-                        if (closed) Pull.done
-                        else Pull.eval(waiting.get) >> consumeLoop
-                      )
+                      Applicative[F].unit
                     }
+                  }
                 }
-                .uncancelable
-            }
-          }.flatten
+              }
 
-        def notifyStream(waitForChanges: Option[Deferred[F, Unit]]) =
-          waitForChanges.traverse(_.complete(()))
+            isClosed.ifM(
+              LeftClosedF,
+              permit(isClosed.ifM(LeftClosedF, q.offer(a.asInstanceOf[AnyRef]).as(RightUnit)))
+            )
+          }
 
-        def waitOnBound(producer: Deferred[F, Unit], poll: Poll[F]) =
-          poll(producer.get).onCancel {
-            state.update { s =>
-              s.copy(producers = s.producers.filter(_._2 ne producer))
+          def trySend(a: A): F[Either[Channel.Closed, Boolean]] =
+            isClosed.flatMap { b =>
+              if (b)
+                LeftClosedF.asInstanceOf[F[Either[Channel.Closed, Boolean]]]
+              else
+                q.tryOffer(a.asInstanceOf[AnyRef]).map(_.asRight[Channel.Closed])
             }
-          }
 
-        def signalClosure = closedGate.complete(())
+          val stream: Stream[F, A] = {
+            val takeN: F[Chunk[A]] =
+              q.tryTakeN(None).flatMap {
+                case Nil =>
+                  // if we land here, it either means we're consuming faster than producing
+                  // or it means we're actually closed and we need to shut down
+                  // this is the unhappy path either way
+
+                  val fallback = q.take.flatMap { a =>
+                    // if we get the sentinel, shut down all the things, otherwise emit
+                    if (a eq Sentinel)
+                      closedLatch.complete(()).as(Chunk.empty[A])
+                    else
+                      Chunk.singleton(a.asInstanceOf[A]).pure[F]
+                  }
+
+                  // check to see if we're closed and done processing
+                  // if we're all done, complete the latch and terminate the stream
+                  isQuiesced.flatMap { b =>
+                    if (b)
+                      closedLatch.complete(()).as(Chunk.empty[A])
+                    else
+                      fallback
+                  }
+
+                case as =>
+                  // this is the happy path: we were able to take a chunk
+                  // meaning we're producing as fast or faster than we're consuming
+
+                  isClosed.flatMap { b =>
+                    val back = if (b) {
```

Comment:
the detection and stripping of the Sentinel seems a bit obscure, what's the rationale? Don't we have the guarantee that the Sentinel, if present, is the last element of this Chunk?

Reply:
> Don't we have the guarantee that the Sentinel, if present, is the last element of this Chunk?

We do, but if the `Chunk` is a `seq` then its contents will be a `List`, so we can't look at the last element efficiently. Though, thinking about it, we can probably pull together a recursive function that is strictly faster than `filter` since we *only* have to look at the last element in this case…